interfaces/builtin/compatibility.go

```diff
@@ -0,0 +1,180 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// CompatRange is a range of unsigned integers, with Min <= Max.
+type CompatRange struct {
+	Min uint
+	Max uint
+}
+
+// CompatDimension represents a <tag>[-<integer|integer range>] (dimension) in a compatibility field.
+type CompatDimension struct {
+	// Tag is the dimension identifier
+	Tag string
+	// Values is the list of integer/integer ranges. Only the last range
+	// can have Min!=Max.
+	Values []CompatRange
+}
+
+type CompatField struct {
+	Dimensions []CompatDimension
+}
+
+const (
+	maxDimensions = 3
+	maxIntegers   = 3
+)
+
+var (
+	tagRegexp = regexp.MustCompile(`^[a-z][a-z0-9]{0,31}$`)
+	// 0 or a number that starts with 1-9 with a maximum of 8 digits
+	intRegexp      = regexp.MustCompile(`^(0|[1-9][0-9]{0,7})$`)
+	intRangeRegexp = regexp.MustCompile(`^\((0|[1-9][0-9]{0,7})\.\.(0|[1-9][0-9]{0,7})\)$`)
+)
+
+func isRealCompatRange(cr CompatRange) bool {
+	return cr.Min < cr.Max
+}
+
+func appendDimension(dimensions []CompatDimension, dimension *CompatDimension) []CompatDimension {
+	if dimension != nil {
+		if len(dimension.Values) == 0 {
+			// If we have just a tag, that is equivalent to <tag>-0
+			dimension.Values = []CompatRange{{0, 0}}
+		}
+		dimensions = append(dimensions, *dimension)
+	}
+	return dimensions
+}
+
+func appendRange(ranges []CompatRange, rg CompatRange) ([]CompatRange, error) {
+	numVal := len(ranges)
+	if numVal > 0 && isRealCompatRange(ranges[numVal-1]) {
+		return nil, fmt.Errorf("ranges only allowed at the end of compatibility field")
+	}
+	if numVal == maxIntegers {
+		return nil, fmt.Errorf("only %d integer/integer ranges allowed per dimension in compatibility field", maxIntegers)
+	}
+	return append(ranges, rg), nil
+}
+
+func parseIntegerRange(token, compat string) (rg *CompatRange, err error) {
+	if intRegexp.MatchString(token) {
+		// 8 decimals fit 32 bits
+		min, err := strconv.ParseUint(token, 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		// Integers are stored as ranges where min==max
+		rg = &CompatRange{uint(min), uint(min)}
+	} else {
+		matches := intRangeRegexp.FindStringSubmatch(token)
+		// Not a tag, integer or integer range
+		if len(matches) == 0 {
+			return nil, fmt.Errorf("invalid tag %q in compatibility field %q",
+				token, compat)
+		}
+		// 8 decimals fit 32 bits
+		min, err := strconv.ParseUint(matches[1], 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		max, err := strconv.ParseUint(matches[2], 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		if min > max {
```

Comment:
in general does it make sense to support writing (a..a) ?

Reply:
I think it is ok to allow this, the meaning is clear and maybe if somebody automatically generates interface definition it might make their life easier.